app/src/main/java/bg/zahov/app/ui/workout/start/StartWorkoutViewModel.kt

```diff
@@ -1,117 +1,229 @@
+     */
     init {
-        getWorkouts()
         viewModelScope.launch {
-            workoutState.state.collect {
-                when (it) {
-                    WorkoutState.MINIMIZED -> _state.postValue(State.Active(true))
-                    WorkoutState.INACTIVE -> _state.postValue(State.Active(false))
-                    else -> {}
+            launch {
+                try {
+                    repo.getStartWorkouts().collect { workouts ->
+                        _uiState.update { old ->
+                            old.copy(workouts = workouts)
+                        }
+                    }
+                } catch (e: CriticalDataNullException) {
+                    serviceError.initiateCountdown()
+                }
+            }
+            launch {
+                try {
+                    repo.getTemplateExercises().collect { templates ->
+                        exerciseTemplates = templates
+                    }
+                } catch (e: CriticalDataNullException) {
+                    serviceError.initiateCountdown()
+                }
+            }
+            launch {
+                workoutState.state.collect {
+                    currentWorkoutState = it
                 }
             }
         }
     }
 
-    private fun getWorkouts() {
+    /**
+     * Starts an empty workout or one from template if one isn't active
+     * else notifies the user
+     */
+    fun startWorkout(workout: StartWorkout? = null) {
         viewModelScope.launch {
-            try {
-                repo.getTemplateWorkouts().collect {
-                    _templates.postValue(it)
-                }
-            } catch (e: CriticalDataNullException) {
-                serviceError.initiateCountdown()
+            if (currentWorkoutState == WorkoutState.INACTIVE) {
+                workoutState.startWorkout(workout = workout?.toWorkout(exerciseTemplates))
+            } else {
+                toastManager.showToast(R.string.starting_workout_while_another_is_active)
             }
         }
     }
 
-    fun startEmptyWorkout() {
+    /**
+     * Adds a duplicate template workout to the repository.
+     *
+     * This method creates a new `Workout` object based on the `StartWorkout` passed as input, appending the word "duplicate"
+     * to the name if the workout's name does not already contain the word "duplicate". The newly created workout is then
+     * added to the repository as a template workout.
+     *
+     * @param newWorkout The `StartWorkout` object that contains the details of the workout to duplicate.
+     */
+    fun addDuplicateTemplateWorkout(newWorkout: StartWorkout) {
         viewModelScope.launch {
-            workoutState.startWorkout(null)
+            repo.addTemplateWorkout(
+                Workout(
+                    id = generateRandomId(),
+                    name = if (newWorkout.name.contains("duplicate")) newWorkout.name else "${newWorkout.name} duplicate",
```

Comment:
Looks like a string that needs extracting.

Reply:
👍